### ISSUE TITLE
Tracking issue for RFC PR 385 - module system cleanup

### ISSUE THREAD

**Issue by nrc** [rust-lang/rust#18219]
* Remove the ordering requirement for `mod`, `extern crate`, and other items;
* allow `extern crate` in non-module blocks;
* allow `pub extern crate`.

**Comment by aturon** [comment]
Nominating. From previous discussion it's not clear what milestone this belongs on, if any.

**Comment by pnkfelix** [comment]
Removing the ordering restriction would be nice for 1.0 (not 1.0-beta).  The other two bullets are not necessary for 1.0.

1.0 milestone.

**Comment by nrc** [comment]
Note to self: work out what still needs to be done

**Comment by nikomatsakis** [comment]
@nrc 

```rust
#![feature(core)]

pub extern crate core;

fn main() { }
```

compiles on play: http://is.gd/RZrZ5X

**Comment by nrc** [comment]
So, `pub extern crate` appears to work. `extern crate` in non-module blocks only seems to partially work. E.g.,:

```
#![feature(core)]

fn main () {
    extern crate core;
    let _ = core::cell::UnsafeCell::new(42i32);
}
```

Does not give an error. But, `use core::cell;` does not work - use expressions don't seem to be able to see the crate brought into scope.

The ordering restrictions are gone.

I'll close this issue an open one for the extern crate bug.

**Comment by glaebhoerl** [comment]
(which is #24219 for anyone else wondering)
